crates/red_knot_python_semantic/src/types.rs

```diff
@@ -673,6 +673,36 @@ impl<'db> Type<'db> {
             Type::Tuple(_) => builtins_symbol_ty(db, "tuple"),
         }
     }
+
+    /// Return the string representation of this type when converted to string as it would be
+    /// provided by the builtin `str` function at runtime (see `__str`). If that can't be
+    /// determined, return `None`.
+    ///
+    /// When not available, this should fall back to the value of `[Type::repr]`.
+    /// Note: this method is used in the builtins `format`, `print`, `str.format` and `f-strings`.
+    pub fn str(&self, db: &'db dyn Db) -> Option<String> {
+        let str_result = match self {
+            // For those primitive types, `str` and `repr` are equivalent so we rely on `repr` to
+            // avoid code duplication.
+            Type::IntLiteral(_) | Type::BooleanLiteral(_) | Type::StringLiteral(_) => None,
+            // TODO: handle more complex types
+            _ => None,
+        };
+        str_result.or_else(|| self.repr(db))
+    }
+
+    /// Return the string representation of this type as it would be provided by the builtin `repr`
+    /// function at runtime (see `__repr__`). If that can't be determined, return `None`.
+    pub fn repr(&self, db: &'db dyn Db) -> Option<String> {
```

Comment:
I think that both this method and `Type::str` should instead return a `Type<'db>`, which in practice will always be either `Type::LiteralString` or a `Type::StringLiteral`. In some other cases, returning a custom enum allows more flexibility in the caller and sometimes more efficiency, but that's not the case here. Returning a `String` means allocating, and we don't even get to take advantage of type interning; we'll only ever have a single `StringLiteralType("True")` (because of Salsa interning), but we could end up allocating many different "True" Strings from this method as currently written. Also, it means that `Type::str` on a `StringLiteral` can just return the same type, not allocate a new owned String.